File: Julia Libraries/create_chirp.jl
#=  create_chirp()
    This is a function that will create a chirp that can be used for the time domain finder
    algorithm. This uses the chirp function from the DSP package.

    Inputs:     chirp_type::string  ==  linear or log scaling
                freq_start::int     ==  Start frequency of chirp (Hz)
                freq_end::int       ==  End frequeny of chirp (Hz)
                fs::int             ==  Sample frequency from test (Hz)
                show_chirp::bool    ==  option to show chirp
                show_spect::bool    ==  option to show spectrogram
                number::int         ==  number of chirps per second

    Sam Kramer
    September 5th, 2023
=#

# --Include statements


# --Using Statements
    using SignalAnalysis
    using GLMakie
    using DSP 


# --Function definition
function create_chirp(freq_start, freq_end, duration, fs, show_response)

# --Generate chirp signal
    chirp_signal = chirp(freq_start, freq_end, duration, fs);   # Call chirp function
    chirp_signal = convert(Vector, chirp_signal);               # convert from MetaArray into a complex vector
    chirp_signal = real(chirp_signal);                          # Convert into a real array.

# --Show plots of signals, will default to false

        if show_response == "true"
            t = vec(0:1/fs:duration - (1/fs));

            scene1 = lines(t, chirp_signal, xlabel = "Time (s)", ylabel = "Amplitude", title = "Generated Chirp Signal");

            spec = spectrogram(chirp_signal, 300, 290, fs = fs);
            y = Plots.heatmap(spec.time, spec.freq, spec.power, xlabel = "Time (s)", ylabel = "Frequency", title = "Spectrogram", color = :jet)
            display(y)

        end

    return chirp_signal
end

File: Libraries/Matlab Libraries/create_chirp.m
function signal = create_chirp(chirp_type, number, length, freq_start, freq_end, fs, show_spect)
%{
    This function will be used to create a chirp signal that will be used
    for the advanced filtering of the signals that are returned from the
    bat recordings.

    Inputs:     chirp_type  == linear or logarithmic, takes a string
                freq_start  == Start frequency of chirp  (Hz)
                freq_end    == Ending frequency of chirp (Hz)
                fs          == Sample frequency (Hz)
                show_chirp  == option to show chirp boolean
                show_spect  == option to show spectrogram bool
                number      == number of chirps per second
                

    Sam Kramer
    July 11th, 2023
%}

% --Call on the dsp.chirp function
    number = length/number;
    chirp = dsp.Chirp('SweepDirection', 'Unidirectional', ...
                      'Type', chirp_type, ...
                      'TargetFrequency', freq_end, ...
                      'InitialFrequency', freq_start,...
                      'TargetTime', number, ...
                      'SweepTime', number, ...
                      'SamplesPerFrame', length*fs, ...
                      'SampleRate', fs);

% --Output Signal
    signal = chirp();

% --Show Spectrogram
    if show_spect == "True" || show_spect == "true"
        try
            spectrogram_function(signal,fs,0,length)   % My own function

        catch
            [s,f,t] = spectrogram(signal, hamming(300), 290, [], fs,'yaxis');
                t = t.*1000;
                f = f ./ 1000;
                s = abs(s);
                s = s-min(min(s));
                s = s ./ max(max(s));
                imagesc(t,f,s)
                set(gca,"YDir","normal")
                colormap('jet')
                clb = colorbar;
                % clim([-60 0])
                title('Unfiltered Spectrogram of Data')
                xlabel('Time (ms)');
                ylabel('Frequency (kHz)')
                clb.Label.String = "Normalized Amplitude";
                clb.Label.set("FontSize",12)
                clb.Label.set("FontName","Serif")
                maxf = max(max(f));
                minf = min(min(f));
                ylim([minf maxf])

        end
    end
end